src/sentry/sentry_metrics/multiprocess.py

```diff
@@ -0,0 +1,334 @@
+import functools
+import logging
+import time
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generic,
+    List,
+    Mapping,
+    MutableMapping,
+    Optional,
+    Sequence,
+    TypeVar,
+)
+
+from arroyo.backends.kafka import KafkaConsumer, KafkaPayload, KafkaProducer
+from arroyo.processing import StreamProcessor
+from arroyo.processing.strategies import ProcessingStrategy
+from arroyo.processing.strategies import ProcessingStrategy as ProcessingStep
+from arroyo.processing.strategies import ProcessingStrategyFactory
+from arroyo.processing.strategies.streaming.transform import ParallelTransformStep
+from arroyo.types import Message, Partition, Position, Topic
+from django.conf import settings
+
+from sentry.utils import json, kafka_config
+
+TPayload = TypeVar("TPayload")
+TProcessed = TypeVar("TProcessed")
+
+DEFAULT_QUEUED_MAX_MESSAGE_KBYTES = 50000
+DEFAULT_QUEUED_MIN_MESSAGES = 10000
+
+logger = logging.getLogger(__name__)
+
+
+def initializer():
+    from sentry.runner import configure
+
+    configure()
+
+
+def get_config(topic: str, **options) -> MutableMapping[str, Any]:
+    consumer_config = kafka_config.get_kafka_consumer_cluster_options(
+        "default",
+        override_params={
+            "enable.auto.commit": False,
+            "enable.auto.offset.store": False,
+            "group.id": "ingest-metrics-consumer",
+            "default.topic.config": {"auto.offset.reset": "latest"},
+            # overridden to reduce memory usage when there's a large backlog
+            "queued.max.messages.kbytes": DEFAULT_QUEUED_MAX_MESSAGE_KBYTES,
+            "queued.min.messages": DEFAULT_QUEUED_MIN_MESSAGES,
+        },
+    )
+    return consumer_config
+
+
+class MetricsBatchBuilder(Generic[TPayload]):
+    def __init__(self, max_batch_size: int, max_batch_time: float) -> None:
+        self.__messages = []
+        self.__max_batch_size = max_batch_size
+        self.__deadline = time.time() + max_batch_time
+
+    def __len__(self) -> int:
+        return len(self.__messages)
+
+    @property
+    def messages(self):
+        return self.__messages
+
+    def append(self, message: Message[TPayload]) -> None:
+        self.__messages.append(message)
+
+    def ready(self) -> bool:
+        if len(self.messages) >= self.__max_batch_size:
+            return True
+        elif time.time() >= self.__deadline:
+            return True
+        else:
+            return False
+
+
+class BatchMessages(ProcessingStep):
```

Comment:
You can still make this strictly typed by using generics:
```
class BatchMessages(ProcessingStep[TPayload]):
...
# then you will have to do this 
def submit(self, message: TPayload) -> None:
    ....
```

And mypy will know what the type of message is that you expect in submit and will be able to tell you that you are misusing the class if you initialize a `BatchMessages(ProcessingStep[KafkaMEssage])` and pass an integer to the submit call.

This is the advantage of building generic logic.